maintenance/ApplyRules.php

```diff
@@ -19,7 +22,30 @@ public function __construct() {
 	}
 
 	public function execute() {
-		// TODO: Implement execute() method.
+		$services = MediaWikiServices::getInstance();
+		$pageRecords = $services->getPageStore()->newSelectQueryBuilder()->fetchPageRecords();
+		$wikiPageFactory = $services->getWikiPageFactory();
+		$pageCount = 0;
+
+		foreach ( $pageRecords as $pageRecord ) {
+			/** @var PageRecord $pageRecord */
+			$page = $wikiPageFactory->newFromTitle( $pageRecord );
+			$categoryCountBefore = count( $page->getCategories() );
+
+			// Get ParserOutput without cache to trigger onContentAlterParserOutput
+			$parserOutput = $page->getParserOutput( noCache: true );
+			( new LinksUpdate( $page->getTitle(), $parserOutput ) )->doUpdate();
+			$page->doPurge();
+
+			$titleText = $page->getTitle()->getPrefixedText();
+			$categoryCountAfter = count( $page->getCategories() );
+
+			$this->output( "Applied rules on {$titleText} ($categoryCountBefore -> $categoryCountAfter)\n" );
```

Comment:
The category counts are for partial debugging and should be removed. It's probably not worth the time to log what categories were added/removed.